river/LayoutManager.zig

```diff
@@ -0,0 +1,378 @@
+pub const Layout = struct {
+    layout: *zriver.LayoutV1 = undefined,
+    namespace: ?[]const u8 = null,
+    output: *Output = undefined,
+
+    pub fn init(self: *Layout, namespace: [*:0]const u8, output: *Output, client: *wl.Client, version: u32, id: u32) !void {
+        self.* = .{
+            .output = output,
+            .namespace = try util.gpa.dupe(u8, std.mem.span(namespace)),
+        };
+
+        self.layout = try zriver.LayoutV1.create(client, version, id);
+        self.layout.setHandler(*Layout, handleRequest, null, self);
+
+        // Namespaces must be unique, so if any other layout on this output
+        // already has the requested namespace it can not be chosen by this
+        // one. In that case we send the namespace_in_use event to the client
+        // and free the namespace buffer, marking the Layout object as invalid.
+        var it = output.layouts.first;
+        while (it) |layout_node| : (it = layout_node.next) {
+            if (std.mem.eql(u8, self.namespace.?, layout_node.data.namespace.?)) {
+                self.layout.sendNamespaceInUse();
+                util.gpa.free(self.namespace.?);
+                self.namespace = null;
+                return;
+            }
+        }
+    }
+
+    pub fn deinit(self: *Layout) void {
+        self.layout.destroy();
+        if (self.namespace) |namespace| util.gpa.free(namespace);
+    }
+
+    /// Send a layout demand to the client
+    pub fn startLayoutDemand(self: *Layout, views: u32) void {
+        log.debug(
+            .layout_manager,
+            "starting layout demand '{}' on output '{}'",
+            .{ self.namespace, self.output.wlr_output.name },
+        );
+
+        // First we create the Layoutdemand object
+        self.output.layout_demand = util.gpa.create(LayoutDemand) catch return;
+        self.output.layout_demand.?.init(self.output, views);
+
+        // Then we let the client know that we require a layout
+        self.layout.sendLayoutDemand(
+            views,
+            self.output.main_count,
+            wl.Fixed.fromDouble(self.output.main_factor),
+            self.output.usable_box.width,
+            self.output.usable_box.height,
+            self.output.layout_demand.?.serial,
+        );
+
+        // And finally we advertise all visible views
+        var it = ViewStack(View).iter(self.output.views.first, .forward, self.output.pending.tags, Output.arrangeFilter);
+        while (it.next()) |view| {
+            self.layout.sendAdvertiseView(
+                view.pending.tags,
+                view.getAppId() orelse "[no-appid]",
+                self.output.layout_demand.?.serial,
+            );
+        }
+        self.layout.sendAdvertiseDone(self.output.layout_demand.?.serial);
+    }
+
+    fn handleRequest(layout: *zriver.LayoutV1, request: zriver.LayoutV1.Request, self: *Layout) void {
+        switch (request) {
+            .destroy => {
```

Comment:
This should do nothing but call `layout.destroy()`. The actual destruction should then happen in the destroy function passed to `setHandler()`. See this section of the wlroots contributing guide for other good rules to follow when implementing protocols server-side: https://github.com/swaywm/wlroots/blob/master/CONTRIBUTING.md#wayland-protocol-implementation